securedrop/journalist_app/api2/__init__.py

```diff
@@ -0,0 +1,148 @@
+    each request: e.g., a series of requests with the prefixes ``{0...f}`` will
+    effectively shard the index into 16 shards.
+    """
+    sources = {}
+
+    query = all_sources()
+    if prefix is not None:
+        if len(prefix) >= PREFIX_MAX_LEN:
+            abort(
+                422, f"malformed request; prefix must be shorter than {PREFIX_MAX_LEN} characters"
+            )
+
+        query = query.filter(Source.uuid.startswith(prefix))
+
+    for source in query.all():
+        all_source_metadata = source.to_api_v2()
+        source_info: Dict[str, Any] = {
+            "version": json_version(all_source_metadata["source"]),
+            "collection": {},
+        }
+        for uuid, item in all_source_metadata["collection"].items():
+            source_info["collection"][uuid] = json_version(item)
+        sources[source.uuid] = source_info
+
+    index = {"sources": sources}
+    version = json_version(index)
+    response = jsonify(index)
+
+    # If the request's `If-None-Match` header matches the version,
+    # return HTTP 304 with an empty response.
+    response.set_etag(version)
+    return response.make_conditional(request)
+
+
+@blp.post("/sources")
+def sources() -> Response:
+    """
+    Return the source metadata requested in the source delta.  For "full"
+    sources, return all metadata.  For "partial" sources, return metadata only
+    for the specified items in their collections, excluding the source-level
+    "info" object.
+
+        {
+            "full_sources": [<source_uuid>, ...],
+            "partial_sources": {
+                "<source_uuid>": [<item_uuid>, ...],
+                ...
+                }
+        }
+
+    The client MAY choose an arbitrary source delta with each request, e.g.
+    from a shard retrieved from ``/index/<prefix>``.
+    """
+    # Parse and validate the request body
+    requested = request.json
+    if not isinstance(requested, dict):
+        abort(422, "malformed request")
+    if (
+        "full_sources" not in requested
+        or not isinstance(requested["full_sources"], list)
+        or not all(isinstance(item, str) for item in requested["full_sources"])
+    ):
+        abort(422, "malformed request; full_sources must be a list of strings")
+    if "partial_sources" not in requested or not isinstance(requested["partial_sources"], dict):
+        abort(422, "malformed request; partial_sources must be a dict")
+    for value in requested["partial_sources"].values():
+        if not isinstance(value, list) or not all(isinstance(item, str) for item in value):
+            abort(422, "malformed request; each value in partial_sources must be a list of strings")
+
+    # Look up all requested sources, and return both "full" and "partial"
+    # metadata in a single pass.
+    source_lookup = set(requested["full_sources"]) | set(requested["partial_sources"].keys())
+    response: Dict[str, Dict[str, Any]] = {"sources": {}}
```

Comment:
nit: is this `"source"` key necessary in the response? seems like we could just return
```
{
  "<source_uuid>": {
    "info": {...},
    "collection": [...]
  },
  ..
}
```